content/blog/2025-05-13-bril-concurrency-extension.md

```diff
@@ -0,0 +1,163 @@
+
+=== Benchmark: perf_big_par_sum_10M ===
+Sequential: 15.8 s, Concurrent:  > (timeout/slower) → RPC overhead dominates
+```
+
+#### **Why “concurrent” is actually slower**
+
+1.  Our design (real Web Workers + RPC heap) means _every_ load and store from a worker turns into:
+    ```
+    postMessage({ kind: "heap_req", … })   // worker → main
+    → main services request, touches native JS array
+    postMessage({ kind: "heap_res", … })   // main → worker
+    ```
+ On a 100 k–element loop, that’s ∼200 k round trips, each costing hundreds of microseconds in Deno’s message‐passing layer—totalling several seconds. At 10 M elements, it grows to ∼20 M messages, which simply overwhelms the system.
+    
+2.  Spawning even two workers in Deno carries a fixed overhead (tens to hundreds of milliseconds). In fine-grained loops (100 k or fewer iterations), that overhead is not negligible, and in very fine loops (10 M) it compounds the RPC penalty.
+    
+3.  The main isolate is busy servicing heap requests from _both_ workers. The event loop context‐switching and message queue flooding create contention, so neither the “workers” nor the main thread run at full core capacity.
+    
+I'd expect that this implementation of concurrency would help, on moderately coarse workloads** (e.g. 100 k or splitting a 100 × 100 matrix) still see _some_ parallelism, because the computation per RPC is nontrivial (simple arithmetic plus pointer arithmetic inside V8). In our tests, the sequential run was ~4 s, the concurrent ~45 s—still slower, but less horrific than the 10 M sum case.
+    
+Some lessons/thoughts.  To actually _win_ with real parallelism under this design, we must batch memory operations. For example, transform long loops into single RPC calls that process entire slices (e.g. “sum these 1 000 elements” in one go), amortizing the message‐passing cost.SharedArrayBuffers could eliminate RPC entirely by mapping our Bril heap into a typed array visible in all workers. Then each load/store is a direct memory access, and you’d see true multicore speedups on large-N benchmarks. For an intermediate step, we could group every 1 000 loads/stores into one batched message, cutting messaging overhead by two orders of magnitude, which should already push the breakeven point down toward the 10 M-element range.
+    
+_In summary_, our RPC-based “shared-heap” prototype cleanly demonstrates the mechanics of true concurrency in Bril, but the messaging overhead severely limits speedups for fine-grained loops. Future work on shared‐memory backing or batched RPC will be necessary to turn this into a truly performant parallel interpreter.
+
+### 5.3 Escape Analysis Statistics
+
+```
+[escape] main: 1/1 allocs are thread-local   # solo allocation stays local
+[escape] TOTAL: 1/1 (100%) thread-local
+
+[escape] main: 0/1 allocs are thread-local   # spawn causes escape
+[escape] mixed: 1/2 allocs are thread-local  # one local, one escaping
+```
```

Comment:
Instead of just pasting the output of your tool can you explain what this is saying? What are "main" and "mixed"? Why are the numbers so small (1 and 2)?